### DISCUSSION TITLE
Allow developers to pass in custom messages into ForbiddenException

### DISCUSSION THREAD

**Discussion by cube-dan** [directus/directus#16631] [Feature Requests]
Greets Team Directus, 

I'd love to be able to pass in messages to ForbiddenException, like this:
`throw new ForbiddenException('some message here');`
For example, from a [custom endpoint](https://docs.directus.io/extensions/endpoints.html), like this: 
```
router.get('/protected', expressAsyncHandler( async (req, res, next) => {
   // other code
   if(!can_access){
      throw new ForbiddenException('some message here');
   }
}
```
That way I can control the message that comes back from the API, for example: 
`{"errors":[{"message":"some message here","extensions":{"code":"FORBIDDEN"}}]}`

This request may dovetail into #4368.

**What I'm doing for now** 

Currently I'm using [patch-package](https://www.npmjs.com/package/patch-package) (patch-package:^6.5.0) to hack custom messages into this file: 
`./node_modules/directus/dist/exceptions/forbidden.js` ... but I'd love to not have to do that. 

Here's a copy of my patch file (as .txt ... because github)
[directus+9.21.0.patch.txt](https://github.com/directus/directus/files/10097232/directus%2B9.21.0.patch.txt)

And the full contents of my ./node_modules/directus/dist/exceptions/forbidden.js
```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForbiddenException = void 0;
const exceptions_1 = require("@directus/shared/exceptions");
class ForbiddenException extends exceptions_1.BaseException {
    constructor(message=null) {
        message = new String(message) || `You don't have permission to access this.`;
        super(message, 403, 'FORBIDDEN');
        /**
         * We currently don't show the reason for a forbidden exception in the API output, as that
         * has the potential to leak schema information (eg a "No permission" vs "No permission to files"
         * would leak that a thing called "files" exists.
         * Ref https://github.com/directus/directus/discussions/4368
         */
    }
}
exports.ForbiddenException = ForbiddenException;
```

Thanks for your consideration. 

-Dan

**Comment by cube-dan** [comment]
Updated patch to remove String('null') as default message when no message is passed in. New patched version: 

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForbiddenException = void 0;
const exceptions_1 = require("@directus
[directus+9.21.0.patch.txt](https://github.com/directus/directus/files/10190341/directus%2B9.21.0.patch.txt)
/shared/exceptions");
class ForbiddenException extends exceptions_1.BaseException {
    constructor(message=null) {
        // set a default message if a non-string message was passed in 
        if(typeof message !== 'string'){
          message = `You don't have permission to access this.`;
        }
        super(message, 403, 'FORBIDDEN');
        /**
         * We currently don't show the reason for a forbidden exception in the API output, as that
         * has the potential to leak schema information (eg a "No permission" vs "No permission to files"
         * would leak that a thing called "files" exists.
         * Ref https://github.com/directus/directus/discussions/4368
         */
    }
}
exports.ForbiddenException = ForbiddenException;
```
[directus+9.21.0.new.patch.txt](https://github.com/directus/directus/files/10190350/directus%2B9.21.0.new.patch.txt)

**Comment by rijkvanzanten** [comment]
Even though it's way more user friendly, I'm a little worried about adding messages to the forbidden exception for the reason outlined in the comment in the linked file:

```
         * We currently don't show the reason for a forbidden exception in the API output, as that
         * has the potential to leak schema information (eg a "No permission" vs "No permission to files"
         * would leak that a thing called "files" exists.
```

**Comment by cube-dan** [comment]
Thanks for the response @rijkvanzanten.

Within a custom api endpoint, what would be the best way to deny access—**with a custom error message**—since I shouldn't be able to do this using the builtin `ForbiddenException`?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
